[#6089] Fix typeahead searches

gbp opened the pull request:
## Relevant issue(s)

Fixes #6089

## What does this do?

Update expected exception message when wildcard searches error allowing
fallback to an non-wildcard search to continue.

## Why was this needed?

Poor search performance

## Implementation notes

Locally in dev and test env all was working fine. I could only replicate in production, this is due to the size of the index. A large index can't always do a wildcard search, for example, when searching for `Cabinet Office` it will raise an exception with the message like: 
`WildcardError: Wildcard office* expands to more than 1000 terms`

This message seems to have changed at some point when upgrading Xapian in the past.

gbp commented:
It appears`QueryParserError` still exists in the Xapian so maybe we need to catch both `QueryParserError` and `WildcardError`

gbp commented:
Exception message matching was first added in https://github.com/mysociety/alaveteli/commit/22b2db48c005be1d1b87dd6b58f830a2a74d2108

garethrees reviewed (approved):
Makes sense in principle!

Would be worth expanding the details into the commit message, and ideally link to the upstream code where these errors exist so that we have a better future reference.

> It appears `QueryParserError` still exists in the Xapian so maybe we need to catch both `QueryParserError` and `WildcardError`

Yeah maybe they split these into two distinct concepts. [We reference `QueryParserError` elsewhere](https://github.com/mysociety/alaveteli/blob/ad8a2db3e30e28ffea1b9692fa4ab6e8c16b7141/spec/controllers/general_controller_spec.rb#L422) so I assume both still exist, but whether we need to rescue from both here IDK.

> I could only replicate in production

I think it's worth trying to test this fix against the prod DB somehow. Worst case just temporarily patch the live file and run the query in the console.

gbp commented:
@garethrees I've made some changes:

1. managed to replicate this error in a spec by adding a configuration option to `ActsAsXapian`. So this is now tested.
2. moved the `RuntimeError` rescue into `ActsAsXapian` and now re-raising as exception within the `ActsAsXapian` model to allow us to catch know exception classes in the core Alaveteli code.
3. added exception notifications for us

gbp commented:
In trying to write a spec to show the original `QueryParserError` exception message I've worked out that this only is raised when using the `Xapian::QueryParser::FLAG_BOOLEAN` flag which we don't use in typeahead search.

```
[1] pry(main)> ActsAsXapian.readable_init
=> snip
[2] pry(main)> ActsAsXapian.query_parser.parse_query('dog AND', Xapian::QueryParser::FLAG_BOOLEAN)
RuntimeError: QueryParserError: Syntax: <expression> AND <expression>
from (pry):2:in `parse_query'
[3] pry(main)> ActsAsXapian.query_parser.parse_query('dog AND', Xapian::QueryParser::FLAG_LOVEHATE | Xapian::QueryParser::FLAG_SPELLING_CORRECTION | Xapian::QueryParser::FLAG_WILDCARD)
=> #<Xapian::Query:0x00007fd9082652d0 @__swigtype__="_p_Xapian__Query">
```

gbp commented:
In production we are hitting the wildcard expansion limit, which is currently set to 1000. We should consider increasing this limit. With can do this with 9d57051 by adding a `config/xapian.yml` file, something like:

```
test:
  max_wildcard_expansion: 10

development:
  max_wildcard_expansion: 500

production:
  max_wildcard_expansion: 1000
```

From my tests in the production Rails console, increasing to 1075 allows the wildcard query for 'Cabinet Office' to succeed. We would also need to consider server memory usage /cc @sagepe

garethrees commented on lib/acts_as_xapian/acts_as_xapian.rb:
```diff
@@ -42,8 +42,8 @@ module ActsAsXapian
-  class NoXapianRubyBindingsError < StandardError
-  end
+  NoXapianRubyBindingsError = Class.new(StandardError)
+  UnhandleRuntimeError = Class.new(StandardError)
```
```diff
- UnhandleRuntimeError
+ UnhandledRuntimeError
```

garethrees commented on lib/typeahead_search.rb:
```diff
@@ -20,13 +20,11 @@ def xapian_search
-      end
+    rescue ActsAsXapian::UnhandleRuntimeError => e
+      # Wildcard expands to too many terms
+      Rails.logger.info "Wildcard query '#{query}' caused: #{e.message.force_encoding('UTF-8')}"
```
Wondering whether this should be `warn` rather than `info`? Not that it would make much difference to our visibility at the moment, but maybe it might be a way to keep more of an eye on this type of problem that only shows itself over time.

garethrees commented on lib/typeahead_search.rb:
```diff
@@ -21,8 +23,10 @@ def xapian_search
       xapian_search = run_query
     rescue ActsAsXapian::UnhandleRuntimeError => e
-      # Wildcard expands to too many terms
-      Rails.logger.info "Wildcard query '#{query}' caused: #{e.message.force_encoding('UTF-8')}"
```
I think we should keep the log output as well as the optional exception notification.

garethrees reviewed (approved):
LGTM! Amazing work 🦸‍♂️ 

Couple of tiny changes suggested but no need for re-review.
